src/platform/react-hooks/src/ChannelProvider.tsx

```diff
@@ -0,0 +1,68 @@
+import React, { useLayoutEffect, useEffect } from 'react';
+import * as Ably from 'ably';
+import { getContext } from './AblyProvider.js';
+import * as Utils from './utils/utils.js';
+import { channelOptionsWithAgent } from './AblyReactHooks.js';
+
+interface ChannelProviderProps {
+  id?: string;
+  channelName: string;
+  options?: Ably.Types.ChannelOptions;
+  deriveOptions?: Ably.Types.DeriveOptions;
+  children?: React.ReactNode | React.ReactNode[] | null;
+}
+
+export const ChannelProvider = ({
+  id = 'default',
+  channelName,
+  options,
+  deriveOptions,
+  children,
+}: ChannelProviderProps) => {
+  const { client, _channelToOptions } = React.useContext(getContext(id));
+
+  const channel = deriveOptions
+    ? client.channels.getDerived(channelName, deriveOptions)
+    : client.channels.get(channelName);
+
+  useLayoutEffect(() => {
+    if (_channelToOptions[channel.name]) {
+      throw new Error('You can not use more than one `ChannelProvider` with the same channel name');
+    }
+    channel.setOptions(channelOptionsWithAgent(options));
+    _channelToOptions[channel.name] = options ?? {};
+    return () => {
+      delete _channelToOptions[channel.name];
+    };
+  }, [channel, options, _channelToOptions]);
+
+  useEffect(() => {
+    const handleChannelAttached = () => {
+      if (hasOptionsChanged(channel, options)) {
+        throw new Error('You can not use `RealtimeChannel.setOption()` along with `ChannelProvider`');
+      }
+    };
+    channel.on('attached', handleChannelAttached);
+    return () => {
+      channel.off('attached', handleChannelAttached);
+    };
+  }, [channel, options]);
+
+  return <React.Fragment>{children}</React.Fragment>;
+};
+
+const hasOptionsChanged = (channel: Ably.Types.RealtimeChannelPromise, options?: Ably.Types.ChannelOptions) => {
```

Comment:
I don't like that it's basically copy-paste, but I don't want to make this thing public in ably-js

Reply:
Yeah I agree - is there a reason we still can't share code between ably-js core and react-hooks in the v2 branch? I think the main reason we couldn't do this in v2 was to do with us not being able to support es modules but that shouldn't stop us now

Reply:
I decide to remove this check for now, when I tested it there were a lot of corner cases (e.g. modes from server are always in lower-case, default modes for derivedChannel is `['subscribe']`, etc.)